### PR TITLE
[CR]Projectile embedding balancing

### LINKED ISSUE #13165: Embedding weapons in small creatures
I threw my forked spear at a possum and the spear got embedded. And then the possum ran away!

Funny but not really fun or realistic. Recommend some balancing variation like:
* embed chance is random
* examine interaction to remove weapon from creature prior to death
* projectiles bigger than the target cannot embed - OR - if projectile is heavier than the creature, pin it in place?
* variation of 'pinning': creatures with embedded projectile suffer forced weight and volume encumbrance of the projectile. That could model slowing down/pinning small targets. 
* embedded projectiles causing bleeding and/or continuous damage while embedded - bonus if this leaves a blood trail. This would make spear hunting wild game pretty fun.

Attn: @Coolthulhu (who I think created this feature?) for discussion.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/25313292-embedding-weapons-in-small-creatures?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Coolthulhu**
Closes #13165
- Tiny creatures can only have 0 volume items embedded
- Small creatures can only have 0 and 1 volume items embedded
- Added a random chance not to embed, depends on dealt damage (high damage attacks will still always embed)
- High-volume items need higher damage to embed - 3 points per unit of volume

### CODE REVIEW

**Comment by Barhandar** [comment]
Still needs tiny creatures exploding if a projectile above their volume tries to embed.

**Comment by Coolthulhu** [comment]
Exploding happens when the creature dies with low enough negative hp, no need to special case it here.

**Comment by Barhandar** [comment]
Or at least die instantly even if projectile didn't do enough damage.

**Comment by Coolthulhu** [comment]
Instant dying would probably be solved better by making low-accuracy hits miss small creatures rather than resulting in glancing hits. That way all the hits would be relatively high-damage ones.

**Comment by i2amroy** [comment]
Honestly I'm not seeing why just glancing a squirrel with something large should instantly kill it. In truth a glancing blow on a small creature with a large object works almost identically to a glancing blow on a large creature with a small object. The only difference is that while a large creature essentially makes the projectile bounce off of it, hitting a small creature with a large object makes the creature bounce off the object.

For example if we think of the path of a large rock flying through the air, it's path will be relatively unchanged by hitting a squirrel. It's also still totally possible to clip the squirrel without any real damage, it's just that when you do so rather than the rock just bouncing away or glancing off it's the squirrel that gets "glanced" while the rock continues it's path relatively unchanged.

I'd advocate having large projectile make whatever they hit get flung over a square when they "glance" before I'd advocate insta-killing when hit by large projectiles.

That said I'd say this looks pretty okay for now, though I haven't done any extensive testing yet.

**Comment by pipehat** [comment]
Working ok for me so far (applied locally). I would love to see bleeding and ability to pull embedded weapon out of monster. But that could be a separate enhancement.

File: src/ranged.cpp [hunk 1]
```diff
@@ -1536,14 +1536,34 @@ void drop_or_embed_projectile( const dealt_projectile_attack &attack )
...
-    if( mon == nullptr || mon->is_dead_state() ||
-        ( attack.dealt_dam.type_damage( DT_STAB ) +
-          attack.dealt_dam.type_damage( DT_CUT ) <=
-            attack.dealt_dam.type_damage( DT_BASH ) ) ||
-        effects.count( "NO_EMBED" ) != 0 ) {
+
+    // We can only embed in monsters
+    bool embed = mon != nullptr && !mon->is_dead_state();
+    // And if we actually want to embed
+    embed = embed && effects.count( "NO_EMBED" ) == 0;
```
Comment (kevingranade) [inline src/ranged.cpp]: I think you want ```embed &&= effects.count( "NO_EMBED" ) == 0;```

Comment (Coolthulhu) [inline src/ranged.cpp]: Yes, I do.
Does C++ have that operator? My compiler doesn't recognize it.

Comment (kevingranade) [inline src/ranged.cpp]: Bummer, I could have sworn there was one, maybe I'm thinking Perl.

File: src/ranged.cpp [hunk 2]
```diff
@@ -1536,14 +1536,34 @@ void drop_or_embed_projectile( const dealt_projectile_attack &attack )
...
+    // Don't embed in small creatures
+    if( embed ) {
+        const m_size critter_size = mon->get_size();
+        const int vol = dropped_item.volume( true, false );
```
Comment (BevapDin) [inline src/ranged.cpp]: Why `unit_value = true`?

Reading `item::volume`, it seems the `unit_value` gives you the volume of a default sized-stack, not of the actual stack it's called on.

**Comment by BevapDin** [comment]
Looks good and works fine, but the volume thing needs confirmation that it is intended.

File: src/ranged.cpp (hunk 1 above)
Comment (vache) [inline src/ranged.cpp]: Doesn't &= work for this?  I know it's supposed to be for bitwise operations, but I'm sure I've used it in the past for this, and it's worked.

File: src/ranged.cpp (hunk 2 above)
Comment (Coolthulhu) [inline src/ranged.cpp]: I might be reading it wrong, but the only reference to `unit_value` in `item::volume` I see is:
```
if ( unit_value == false ) {
    ret *= charges;
}
ret /= type->stack_size;
```
So for `unit_value == false` the volume should be of a single item of the stack rather than the whole stack.

Comment (BevapDin) [inline src/ranged.cpp]: Yes, but you already have a single item (and if it's counted by charges, it will already have 1 charges only, see end of `game::plthrow`). I have to agree here, `charges` will be 1 and `ret *= charges` will not do anything.

But it's a more formal matter, it's confusing because in theory, it could return less volume explicitly calling with `true` is not needed. It makes the impression that the item *may* have more charges, which btw. would be added to the monster and later dropped anyway.

So in short: if I somehow manage to throw a stack of 100 throwing knifes (should not be possible, but who knows), this function will use the volume of 1 knife, but will add 100 knifes to the monster inventory.

Comment (Coolthulhu) [inline src/ranged.cpp]: I'd say a stack of 100 knives should stack like 100 knives rather than like a stack.

Comment (BevapDin) [inline src/ranged.cpp]: I see now, you're explicitly testing for one item, even if it's a stack because each item is assumed to be embedded separately. The knifes are thrown as a stack, but arrive separately and therefor embed separately, but the code "handles" them as one for it's easier, except here where it needs to check a single item. It's still confusing but it's correct.

Except: the items of the stack either embed all together or none at all. However, this is currently not relevant as the player can not throw a stack of items anyway.
